tracing-subscriber/src/fmt/writer.rs

```diff
@@ -54,6 +54,31 @@ where
     }
 }
 
+/// Prints text using the standard library's macro `print!`. This writer allows stdout logs to be
+/// correctly captured by commands such as `cargo test`.
+#[derive(Debug)]
+pub struct TestWriter;
+
+impl io::Write for TestWriter {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let out_str = String::from_utf8_lossy(buf);
+        print!("{}", out_str);
+        Ok(out_str.len())
```

Comment:
Is `out_str.len()` always equal to `buf.len()`?
If not, output will be slightly broken.

Reply:
I haven't had any issues with broken output so far in my local testing.

I can see how that is a concern given the use of `from_utf8_lossy`. However, the `fmt::Layer` impl passes a String as a byte buffer to the Write trait, so there should never be any loss when going back to a string. Perhaps future-proofing it for just in case can return `buf.len()`.

Reply:
It would probably be fine to use `str::from_utf8` here, and `expect` that it will always be a valid UTF8 string, since the output is guaranteed to come from a valid string in the `fmt` subscriber.  Or, we could just return an `io::Error` in the case that the buffer is not valid UTF-8, if we would prefer not to panic.

Using `str::from_utf8` rather than `String::from_utf8_lossy` also has the advantage of not allocating to create a `String`, so it would give us better performance here. That isn't _too_ important, if we expect this to only ever be used in tests, but it's worth considering...

Reply:
`from_utf8_lossy` returns a `Cow` to avoid allocations on valid UTF-8. Using `from_utf8` requires an additional copy into a `Vec` to be usable.

The `on_event` function in `fmt_layer` ignores the result of `write_all`. I feel a panic is better than dropping messages entirely. `from_utf8_lossy` feels like the sweet spot between dropping messages and a panic. It has the added benefit of being able to see the character `U+FFFD` in the output rather than outright panicking or dropping messages.

Reply:
ah, yup, i forgot it returned a `Cow`. this is right, then!